### ISSUE TITLE
SPFX Angular2+ polyfill support

### ISSUE THREAD

**Issue by ArcanDotNet** [SharePoint/sp-dev-docs#888]
Thank you for reporting an issue or suggesting an enhancement. We appreciate your feedback - to help the team to understand your needs, please complete the below template to ensure we have the necessary details to assist you. 

#### Category
- [ X] Question
- [ ] Typo
- [ ] Bug
- [ ] Additional article idea

> If you are planning to share a new feature request (enhancement / suggestion), please use SP Dev UserVoice at http://aka.ms/sp-dev-uservoice.

#### Expected or Desired Behavior
Fully functional Angular Dev in Internet Explorer via SPFX.  Any suggestions as to how to integrate the necessary Polyfills into the boilerplate or SPFX in general would be greatly appreciated.  I have tried manually adding Polyfills for each issue, but it's too exhaustive a list and each problem you fix reveals a new one.

#### Observed Behavior
Angular 2+ development in SPFX does not work in Internet Explorer because it lacks the necessary Polyfills.  I am using the Angular SPFX Boilerplate: 

http://www.sebastienlevert.com/2017/07/31/sharepoint-framework-and-angular-introducing-the-spfx-angular-boilerplate/

This works great in every browser except IE.  The ES6-Shim causes several errors and does not work correctly.  In Angular 4 there are Polyfills included with the templates that are referenced as needed for IE dev.  I do not see a way to introduce those files into the Gulp based structure of SPFX.  

#### Steps to Reproduce
Download the boiler plate, run a gulp serve and try to use it in Internet Explorer.

Thanks for your contribution! Sharing is caring.

**Comment by NickSevens** [comment]
Hi @ArcanDotNet did you find any way to fix this? We're facing the same issue, in IE11 indeed.
Thanks

**Comment by ArcanDotNet** [comment]
Not yet. Still working with Microsoft on it.  I'll post something here if a workaround is figured out.

**Comment by ArcanDotNet** [comment]
Leave a comment

**Comment by pgonzal** [comment]
+@VesaJuvonen

**Comment by NickSevens** [comment]
@ArcanDotNet indeed, the Map object inside es6-shim seemed to be the cause with us. Thanks for the pointers!

**Comment by pedro-pedrosa** [comment]
The default (SP-Loader) implementation of the Map object breaks when you use it with other core-js polyfills (such as Array.from), hence the need to use Map from core-js as well.

This work around definitely saved me a lot of time, thank you for investigating.

**Comment by javierordonezcastro** [comment]
Did someone find a solution?

I tried what @ArcanDotNet indicated but it did not work. I suppose that the incompatibility in my case comes from another object different from the "Map"

Thank you very much in advance,
Javier

**Comment by pgonzal** [comment]
Polyfills are *global* and affect all components on the page. Although I understand the motivation here, a third-party component really should not be loading polyfills. Even if it doesn't break today (as seen above), it may break with some future SPFx update or in some other context that wasn't tested (e.g. classic pages or a mobile app). 

The sp-polyfills package was designed to be very conservative for performance reasons, since it is guaranteed to be loaded on every page where SPFx runs. In some cases the footprint may be very important, e.g. because the page is shared with a legacy application's scripts. We even discourage first party developers from loading polyfills because it complicates the test matrix and third-party environment for each page.

If there is a community need for heavyweight polyfills, the right solution would be to create an a separate package (e.g. sp-expensive-polyfills) that a web part can depend on. Then it would be loaded only if such a web part is present on the page. This would be a feature request for SPFx.

That said, if you can find some workaround that seems to work, you're free to use that approach as long as you're okay with the risks involved.

**Comment by pgonzal** [comment]
@patmill fyi

**Comment by ArcanDotNet** [comment]
@pgonzal 
A polyfill is a piece of JavaScript code that is intended to exactly simulate a native, interoperable feature of the web platform, for the purpose of allowing website developers to use modern web features in older browsers. -  https://www.w3.org/2001/tag/doc/polyfills/

You either implement the feature to the spec, or you do not.  The sp-polyfills MAP object does not meet the ES6 standard which is why it does not support 3rd party components.  If it did, these components would simply work without the need for loading additional polyfills.

If performance is an issue, I would highly suggest removing the eval function I pointed out above.  This the single most expensive task JavaScript can perform.

**Comment by pgonzal** [comment]
This GitHub issue seemed to start from the position of "If I do something which is not supported for SPFx, I get an error."  I assume you already understand why SPFx cannot reasonably be expected to support custom polyfills; if not I can give more details.

If you believe the **sp-polyfills** `Map` class has a bug or is failing to implement the ES6 standard, please open a separate issue with repro steps.  The error cited above seems to simply be two different polyfills conflicting with each other in their approach for modifying the global object space.  So we'd want to see repro steps that don't involve trying to load `core-js/es6/map`.

Alternatively, if you'd like to propose specific improvements to **sp-polyfills** to make it more compatible with other libraries, that's good feedback.  It was a long time ago when we decided which libraries to include in **sp-polyfills**, and probably the landscape has changed since them.  The primary design goal -- of keeping the bundle size as small as possible -- still stands, though.  For example the default **core-js** bundle is 83kb minified unzipped, which is pretty huge.  And **es6-shim.min.js** is 56kb.  The entire **sp-polyfills** bundle today is only 32kb minified unzipped.

```
var queue = new Array(1000);
function flush() {
  for (var i = 0; i < len; i += 2) {
    var callback = queue[i];
    var arg = queue[i + 1];

    callback(arg);

    queue[i] = undefined;
    queue[i + 1] = undefined;
  }
  len = 0;
}
```

Is this SPFx code?  If so could you provide a little more context?  I tried searching for "queue" and "flush" in our code base but it turned out a huge number of results.  :-)

**Comment by knalbandianbrightgrove** [comment]
Solution for me was removing `import 'core-js/es6/map';` at all

**Comment by pgonzal** [comment]
Does Angular require *all* of core-js? Or is there a core subset of it that is sufficient for most scenarios? We could measure the size of that.

**Comment by jcangelosi** [comment]
@pgonzal No, Angular does not require all of core-js, the following are the most common:
`import 'core-js/es6/symbol';
import 'core-js/es6/object';
import 'core-js/es7/object';
import 'core-js/es6/function';
import 'core-js/es6/parse-int';
import 'core-js/es6/parse-float';
import 'core-js/es6/number';
import 'core-js/es6/math';
import 'core-js/es6/string';
import 'core-js/es6/date';
import 'core-js/es6/array';
import 'core-js/es6/regexp';
import 'core-js/es6/map';
import 'core-js/es6/set';
import 'core-js/es7/array';


/** IE10 and IE11 requires the following for NgClass support on SVG elements */
// import 'classlist.js';  // Run `npm install --save classlist.js`.
/** Evergreen browsers require these. **/
import 'core-js/es6/reflect';
import 'core-js/es7/reflect';

/** IE requires the following pollyfills to support promises and Angular's httpClient */
import 'es6-promise';
import 'whatwg-fetch';`

What I found is that es6/map, es6/reflect, es7/reflect make the sharepoint app crash in Internet Explorer, when I remove them the app loads but then I get a problem with an ngIf in ng Bootstrap alerts when Angular is initating, this happens only in Internet Explorer, so I think I lack one more pollyfill to make it work.

**Comment by pgonzal** [comment]
Are there individual TypeScript typings for these modules? It might be odd to polyfill all these APIs but have no way for people to use them from TypeScript. (We also wouldn't want turn on *all* the ES6 typings since that would enable IntelliSense for APIs that are NOT polyfilled, which would be confusing for developers.)

**Comment by JuanCangelosi** [comment]
Not that I know of.
On monday will try compiling the code on es5 target and not on 6, to see if I can bypass the situation, if i succeed I will post the polyfills I needed to make it work.

**Comment by jcangelosi** [comment]
Not that I know of.
On monday will try compiling the code on es5 target and not on 6, to see if I can bypass the situation, if i succeed I will post the polyfills I needed to make it work.

**Comment by ArcanDotNet** [comment]
Angular adds the list of polyfills above to any new project you create so I believe they are all needed.  The real issue here is that sp-polyfills is adding Map to the global namespace without fulfilling all the requirements of the Map spec in ES6.

If nothing else I would replace that one polyfill with the CoreJS one, or just remove it altogether.  For long term support I would suggest adding a configuration option to SPFX that allows users to add the entire CoreJS library or each section of it as needed.  CoreJS is pretty much the web standard for Polyfills at this point and any attempt to reinvent this functionality is unwise.  

While I have not used Angular Elements, it will likely require ES6 polyfills as well in IE.  Almost all modern frameworks rely on parts of ES6.  It's a 3 year old standard.  But for the moment Custom Elements, the standard on which Angular Elements is based is unsupported in all MS browsers. (IE and Edge)

By recommending Elements the SPFX docs direct contradict SharePoint in terms of support.  SharePoint is recommended to be run on IE11 which has no ES6 support and will never support the Custom Elements standard.  So until Edge does support this new standard and Microsoft updates the SharePoint docs to recommend Edge as the preferred browser, the dev community has no options for using angular in SPFX if they want to full platform compatibility.

This whole issue would be much simpler if IE was end of life, or if SPFX officially supported Angular 2-X.  For now devs should avoid any version of Angular in SPFX if they want their product to be considered inline with SharePoint standards.

**Comment by ArcanDotNet** [comment]
Exactly what it's missing I couldn't say.  But if you add all of the CoreJS polyfills but Map and load Angular you will get an exception.  If you null out the SP-Polyfill version of Map and load any of the other mainline polyfills this does not happen.  This also never happens in modern browsers like Chrome and Edge where Map is natively supported.  This clearly means to me that the current implementation has an issue or does not meet the spec.  If it did, it would not throw exceptions when a common library like Angular tried to use it.

Debugging the exact issue with the SPFX implementation of Map is a lot to ask.  Reproducing the issue is very easy to achieve and I have documented the steps above.

**Comment by jcangelosi** [comment]
After trying a new dummy project with Angular 6 Elements I can confirm that the problem are the SP-Pollyfills, event in a self contained element the SP Map implementation breaks the app with the out of stack space exception.
I think the easiest solution would be for sharepoint to import all the ES6 polyfills, that would mean bigger packages, but would solve a lot of headaches not only for Angular Devs but also for a lot of other developers who may find this issue when polyfilling their apps.

**Comment by SamiEklund** [comment]
Thanks @ArcanDotNet, you saved me a lot of debugging! I got our solution working by removing both 'core-js/es6/map' and 'core-js/es6/set' polyfills. So it seems polyfill for Set might cause this issue too.

**Comment by ArcanDotNet** [comment]
Polyfills don't just implement functionality, they also replace it if it's wrong.  With so many browsers in the world, each defining subsets of ES6 standards the point of polyfills is to evaluate the JS engine and adjust it as needed to support modern JS.  

@SamiEklund You certainly are determined my friend.

**Comment by SamiEklund** [comment]
@ArcanDotNet  My motivation dropped a bit as we did a workaround that worked for us in this scenario. I still might investigate what is missing from the SP polyfill, bit its not a priority atm.

We simply check if `window.Map` or `window.Set` exist, and if not we require the corejs polyfills. So far we've had no issues with it.
```JavaScript
if (!window.Map) {
    require('core-js/es6/map')
}

if (!window.Set) {
    require('core-js/es6/set')
}
```

**Comment by JuanCangelosi** [comment]
Well last week I got notified that the app stopped working in Internet Explorer even with the workaround @ArcanDotNet found, so I started experimenting and it seems that sharepoint added new pollyfills that break even more the app than before, by trial and error I found a new workaround:

`/** IE9, IE10 and IE11 requires all of the following polyfills. **/
import 'core-js/es6/symbol';
import 'core-js/es6/object';
import 'core-js/es6/function';
import 'core-js/es6/parse-int';
import 'core-js/es6/parse-float';
import 'core-js/es6/number';
import 'core-js/es6/math';
import 'core-js/es6/string';
import 'core-js/es6/date';
import 'core-js/es6/array';
import 'core-js/es6/regexp';
if(WeakMap.toString().indexOf('function WeakMap()') === -1)
{
     WeakMap = undefined;
}
import 'core-js/es6/weak-map';

// Check for native support of Map vs Polyfill
if(Map.toString().indexOf('function Map()') === -1)
{
     Map = undefined;
}
import 'core-js/es6/map';
// Check for native support of Map vs Polyfill
if(Set.toString().indexOf('function Set()') === -1)
{
     Set = undefined;
}
import 'core-js/es6/set';`

This is how my polyfills look now, I removed the reflect polyfills and the app is working perfectly fine, I guess it has been added in an update. It's far from Ideal, but at least it works for now.

**Comment by butchmarshall** [comment]
Which file in spfx webpart package do I include this workaround in?

**Comment by SamiSaves** [comment]
You have to add it where ever you are importing the core-js polyfills. If you are not using the polyfills, but still think you are having this issue you should check if one of your dependencies or their dependencies are using the polyfills.

**Comment by ArcanDotNet** [comment]
This was my issue, and I found the workaround.  It seems extremely hackish to me, but I'm not even in the SharePoint realm anymore so I guess people can do what they want.

I feel the team should address it properly and not just embrace an absolute hack that will actually blank out the proper implementation of ES6 code even in modern browsers.
________________________________

**Comment by mrukas** [comment]
I think this issue should be reopened. This is solution is so hacky, it disturbs me.

**Comment by maliksahil** [comment]
ES6 polyfills isn't something you should reinvent - and SPFx has chosen to reinvent them for whatever reasons. This workaround will not work in IE11. It will however work fine in Chrome/Edge/FFox etc. Anyway, the only real solution here is to either
a - Drop support for IE11
b - Not use Angular or Angular elements
c - SPFx to adopt well accepted community standards for polyfills (i.e. get off the sp-polyfills  bandwagon and get on the es6-polyfills bandwagon like the rest of the world).

I feel the best solution is c .. but I'm not holding my breath.

**Comment by hassanamjad89** [comment]
I don't see get this particular issue in SPFx 1.9.1. But recently I did receive 
"TypeError: Object doesn't support property or method 'from'" exception in IE browser when I was using the super awesome 'Yup' validation module in my code. In order to fix this I had to include the following import.
import "core-js/es6/array";

So far this is the only polyfill I had to manually include. Hope Microsoft includes the above polyfill in newer versions of SPFx.

I will update here if in future I am forced to import additional polyfills manually.
